[Addons] RepositoryUpdater: fix time calculation overflow

mglae opened the pull request:
## Description
<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here. -->
Using (32 bit) int for ms calculation lead to (negative) overflow when more than ~25 days ago. Using ~seconds~ std::chrono increase time span to ~68 years.

## Motivation and context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
When updating a LE13 test installation after a few weeks no addon updates were shown. The log looked odd:
```
2025-02-01 18:02:08.178 T:842     debug <general>: CRepositoryUpdater: closest next update check at 2024-12-17 7:04:17 (in 281096 s)
2025-02-01 18:02:08.178 T:842     debug <general>: CRepositoryUpdater: checking in 281096296 ms
```

## How has this been tested?
<!--- Please describe in detail how you tested your change. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
PR included in LE13 build successfully tested on previous failing Addon33.db

## What is the effect on users?
<!--- Summarize the effect of this change on Kodi end-users. -->
<!--- If the PR does not have a noticeable impact (e.g., if it only changes documentation), -->
<!--- just leave it empty. Put in more detail the bigger the impact is. -->
<!--- This section may be used for automatic creation of release notes. -->
Working addon updates.

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` with no space in all the boxes that apply like this: [X] -->
- [X] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` with no space in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

neo1973 commented:
The proposed change should work, but how about a solution using `std::chrono`?
<details>
<summary>diff</summary>

```diff
diff --git a/xbmc/addons/RepositoryUpdater.cpp b/xbmc/addons/RepositoryUpdater.cpp
index c8c1dcbcc32..cd4a7e6b224 100644
--- a/xbmc/addons/RepositoryUpdater.cpp
+++ b/xbmc/addons/RepositoryUpdater.cpp
@@ -313,6 +313,8 @@ CDateTime CRepositoryUpdater::ClosestNextCheck() const
 
 void CRepositoryUpdater::ScheduleUpdate(UpdateScheduleType scheduleType)
 {
+  using namespace std::chrono;
+
   std::unique_lock<CCriticalSection> lock(m_criticalSection);
   m_timer.Stop(true);
 
@@ -322,14 +324,16 @@ void CRepositoryUpdater::ScheduleUpdate(UpdateScheduleType scheduleType)
   if (!m_addonMgr.HasAddons(AddonType::REPOSITORY))
     return;
 
-  int delta{1};
+  milliseconds delta{1};
   const auto nextCheck = ClosestNextCheck();
   if (nextCheck.IsValid())
   {
-    // Repos were already checked once and we know when to check next
-    delta = std::max(1, (nextCheck - CDateTime::GetCurrentDateTime()).GetSecondsTotal() * 1000);
-    CLog::Log(LOGDEBUG, "CRepositoryUpdater: closest next update check at {} (in {} s)",
-              nextCheck.GetAsLocalizedDateTime(), delta / 1000);
+    // Repos were already checked once and we know when to check next.
+    // delta must be positive and not zero (m_timer.Start() ignores 0 wait time)
+    delta = std::max<milliseconds>(
+        delta, seconds((nextCheck - CDateTime::GetCurrentDateTime()).GetSecondsTotal()));
+    CLog::Log(LOGDEBUG, "CRepositoryUpdater: closest next update check at {} (in {})",
+              nextCheck.GetAsLocalizedDateTime(), duration_cast<seconds>(delta));
   }
 
   if (scheduleType == UpdateScheduleType::Regular)
@@ -337,17 +341,12 @@ void CRepositoryUpdater::ScheduleUpdate(UpdateScheduleType scheduleType)
     // Enforce minimum hold-off time of 1 hour between regular updates - this is especially
     // important to handle all sorts of failure cases (e.g., failure to update the add-on database)
     // that would otherwise lead to an immediate new update attempt and continuous hammering of the servers.
-    delta = std::max(1 * 60 * 60 * 1'000, delta);
-  }
-  else
-  {
-    // delta must be positive and not zero (m_timer.Start() ignores 0 wait time)
-    delta = std::max(1, delta);
+    delta = std::max<milliseconds>(hours(1), delta);
   }
 
-  CLog::Log(LOGDEBUG, "CRepositoryUpdater: checking in {} ms", delta);
+  CLog::Log(LOGDEBUG, "CRepositoryUpdater: checking in {}", delta);
 
-  if (!m_timer.Start(std::chrono::milliseconds(delta)))
+  if (!m_timer.Start(delta))
     CLog::Log(LOGERROR,"CRepositoryUpdater: failed to start timer");
 }
 }
```
</details>

Switching to `std::chrono` is one of our distance goals anyway so this should be future proof.

neo1973 committed:
RepositoryUpdater: use std::chrono to fix time calculation overflow

Using (32 bit) int for ms calculation lead to (negative) overflow when more
than ~25 days ago. Use std::chrono increase time span

mglae commented:
> The proposed change should work, but how about a solution using std::chrono?

Sure, when it does compile :)

Do you recommend updating the backport too?

ksooo commented on xbmc/utils/log.h:
```diff
@@ -33,6 +33,7 @@
 #include <string>
 #include <vector>
 
+#include <fmt/chrono.h>
```
Why this change?

mglae replied:
Was required to get it compiled on Linux x86_64.

But all Windows builds are failing in jenkins with libfmt errors. Therefore I will revert this PR to my minimal fix in a few hours.

ksooo replied:
Problem is most probably the call to CLog::Log with delta now having type ‚milliseconds‘ instead of int

neo1973 replied:
Try putting the `#include <fmt/chrono.h>` in  RepositoryUpdater.cpp, that should make Windows happy.

ksooo replied:
Not only Windows, also macOS (just tested).

mglae replied:
@ksooo Windows errors are in UPNP modules.

@neo1973 Just pushed the untested change. May Jenkins decide.

ksooo commented on xbmc/addons/RepositoryUpdater.cpp:
```diff
@@ -322,32 +326,29 @@ void CRepositoryUpdater::ScheduleUpdate(UpdateScheduleType scheduleType)
+    delta = std::max<milliseconds>(hours(1), delta);
   }
 
-  CLog::Log(LOGDEBUG, "CRepositoryUpdater: checking in {} ms", delta);
+  CLog::Log(LOGDEBUG, "CRepositoryUpdater: checking in {}", delta);
```
Just nitpicking, but I would keep the "ms" so the log reader knows the unit of measurement of the value.

neo1973 replied:
The unit is automatically added by the formatter for `std::chrono` types :slightly_smiling_face:

neo1973 reviewed (approved):
Thanks, looks good to me now :+1: 

Once merged please update the backport to match this PR.

ksooo reviewed (approved):
Thanks.

mglae commented:
Backport is updated.
